### PR TITLE
bunx: specify package with --package

### PR DESCRIPTION
**Pull request by telekid**
### What does this PR do?
This PR adds a `--package` flag to `bunx`. I've tried to keep both the syntax and semantics of `--package` as close to NPM's implementation as possible.

Relevant docs:
 - https://docs.npmjs.com/cli/v10/commands/npm-exec
 - https://docs.npmjs.com/cli/v10/commands/npx

**Note for reviewers**

In earlier versions of this code, no distinction was drawn between the package being run and the binary being invoked. The binary name was derived directly from the package. This PR separates those two concepts.

While reviewing, it's helpful to think of there being two separate modes of operation: implicit package and explicit package. In implicit package mode, the package name is assumed to be the same as the first command argument. In explicit mode, the package name is specified via the `--package` flag. 

Tests that are currently failing appear to be doing so because I moved argument parsing to within `bunx_command.zig`. This is causing `bun x` to capture flags like `--help` that were previously passed along directly to the invoked command. I believe this new behavior is actually more consistent with how `node-exec` behaves, with that caveat that I haven't been able to test the difference between `bunx` and `bun x` because I'm not quite sure how to make dev builds of the former yet. Note that this is a breaking change which is probably avoidable. I'm not sure what Bun's stance on breaking changes is. 

**TL;DR:**
`bun x cowsay --help` previously passed `--help` to `cowsay` command, but now passes `--help` to `bun` command. I need input from the bun team regarding the preferred approach here.

For more discussion, see [npx docs](https://docs.npmjs.com/cli/v10/commands/npx#npx-vs-npm-exec).

**TODO**

 - [ ] `.ReplCommand` relies on `.BunxCommand`. Should I be testing anything specific?
 - [ ] test `--bun`. (Currently it looks like the incorrect bin is used with this flag.)
 - [ ] test `bun x tsc main.ts`

**To test**
 - [x] `bun x cowsay hello # this is existing shortcut behavior` [here](https://github.com/oven-sh/bun/pull/7369/files#diff-8223dcf25255bd59c8022233507cdc78a6cef9a6d9d985f7ee3172ab469289f4R32)
 - [ ] `bun x -- cowsay hello`
 - [x] `bun x --package=cowsay -- cowsay hello` [here](https://github.com/oven-sh/bun/pull/7369/files#diff-8223dcf25255bd59c8022233507cdc78a6cef9a6d9d985f7ee3172ab469289f4R251)
 - [x] `bun x --package=cowsay -- cowthink hello` [here](https://github.com/oven-sh/bun/pull/7369/files#diff-8223dcf25255bd59c8022233507cdc78a6cef9a6d9d985f7ee3172ab469289f4R270)
 - [x] `bun x --package=cowsay -- cowthink --version` [here](https://github.com/oven-sh/bun/pull/7369/files#diff-8223dcf25255bd59c8022233507cdc78a6cef9a6d9d985f7ee3172ab469289f4R288)
 - [x] `bun x --package=cowsay hello` errors [here](https://github.com/oven-sh/bun/pull/7369/files#diff-8223dcf25255bd59c8022233507cdc78a6cef9a6d9d985f7ee3172ab469289f4R311)
 - [ ] Test that everything behaves as expected with `bunx` command, too

Closes https://github.com/oven-sh/bun/issues/7034.

### CODE REVIEW

File: src/cli.zig [hunk 1]
```diff
@@ -1343,12 +1338,14 @@ pub const Command = struct {
                 if (comptime bun.fast_debug_build_mode and bun.fast_debug_build_cmd != .BunxCommand) unreachable;
                 const ctx = try Command.Context.create(allocator, log, .BunxCommand);
 
+                // TODO: This probably isn't right anymore
```
Comment (telekid) [inline src/cli.zig]: rm

File: src/cli.zig [hunk 2]
```diff
@@ -182,11 +182,6 @@ pub const Arguments = struct {
     };
     pub const run_params = run_only_params ++ runtime_params_ ++ transpiler_params_ ++ base_params_;
 
-    const bunx_commands = [_]ParamType{
```
Comment (telekid) [inline src/cli.zig:185]: This constant was never actually referenced anywhere, and in this PR I move `bunx` CLI parsing to the `bunx_command` file.

File: src/cli.zig [hunk 3]
```diff
@@ -1894,12 +1891,14 @@ pub const Command = struct {
...
                         \\Flags:
-                        \\  <cyan>--bun<r>      Force the command to run with Bun instead of Node.js
+                        \\  <cyan>--bun<r>          Force the command to run with Bun instead of Node.js
+                        \\  <cyan>--package<r>      Specify package to install
                         \\
                         \\Examples<d>:<r>
                         \\  <b>bunx prisma migrate<r>
                         \\  <b>bunx prettier foo.js<r>
                         \\  <b>bunx<r> <cyan>--bun<r> <b>vite dev foo.js<r>
+                        \\  <b>bunx<r> <cyan>--package<r> <b>prettier prettier foo.js<r>
```
Comment (telekid) [inline src/cli.zig]: TODO: incorrect example; update

File: src/cli/bunx_command.zig [hunk 4]
```diff
@@ -148,43 +149,101 @@ pub const BunxCommand = struct {
...
+                    try passthrough_list.append(positional);
+                }
             }
         }
 
+        if (cli.package.len != 0 and bin_name[0].len == 0) {
+            const text =
+                \\ <r><red>error<r><d>:<r> --package specified without indicating command<r>
+                \\ Perhaps you meant the following?:
+                \\ bunx --package {s} <b>-- {s}<r>
```
Comment (telekid) [inline src/cli/bunx_command.zig:225]: Imperfect suggested action for two reasons:

1. It doesn't include arguments
2. The user may have run `bun x` rather than `bunx`

File: src/cli/bunx_command.zig [hunk 5]
```diff
@@ -148,43 +149,101 @@ pub const BunxCommand = struct {
...
 
                 if (positional.len == 0) continue;
 
-                if (positional[0] != '-') {
-                    if (!found_subcommand_name) {
-                        found_subcommand_name = true;
-                        if (positional.len == 1 and positional[0] == 'x')
-                            continue;
-                    }
+                if (positional.len == 2 and positional[0] == '-' and positional[1] == '-') {
```
Comment (telekid) [inline src/cli/bunx_command.zig:200]: This logic should be lifted into `clap`.

File: src/cli/bunx_command.zig [hunk 6]
```diff
@@ -148,43 +149,101 @@ pub const BunxCommand = struct {
...
+                    parsing_passthroughs = true;
+                    continue;
                 }
 
-                if (!run_in_bun and !found_subcommand_name) {
-                    if (strings.eqlComptime(positional, "--bun")) {
-                        run_in_bun = true;
-                        continue;
-                    }
+                if (!parsing_passthroughs) {
```
Comment (telekid) [inline src/cli/bunx_command.zig:205]: Note to self: is this redundant?

File: src/cli/bunx_command.zig [hunk 7]
```diff
@@ -148,43 +149,101 @@ pub const BunxCommand = struct {
...
                 }
 
-                try passthrough_list.append(positional);
+                if (captured_bin_name and parsing_passthroughs) {
+                    try passthrough_list.append(positional);
+                }
             }
         }
 
+        if (cli.package.len != 0 and bin_name[0].len == 0) {
```
Comment (telekid) [inline src/cli/bunx_command.zig:221]: Specifying `--package` without separating `bunx` flags from `bin` with `--` is ambiguous and confusing, so it seems better to forbid that for now.

File: src/cli/bunx_command.zig [hunk 8]
```diff
@@ -470,4 +522,77 @@ pub const BunxCommand = struct {
...
+
+        // TODO: Finish text
+        const intro_text =
+            \\<b>bunx<r>
+            \\<b>Usage<r>: <b><green>bun install<r> <cyan>[flags]<r> [...\<pkg\>]
+            \\<b>Alias: <b>bun i<r>
+            \\  Install the dependencies listed in package.json
+        ;
+
+        const outro_text =
```
Comment (telekid) [inline src/cli/bunx_command.zig:525]: TODO: Write this text

File: src/cli/bunx_command.zig [hunk 9]
```diff
@@ -148,43 +149,87 @@ pub const BunxCommand = struct {
 
     pub fn exec(ctx_: bun.CLI.Command.Context, argv: [][*:0]const u8) !void {
         var ctx = ctx_;
+
+        const cli = try CommandLineArguments.parse(ctx.allocator);
+
         var requests_buf = bun.PackageManager.UpdateRequest.Array.init(0) catch unreachable;
-        var run_in_bun = ctx.debug.run_in_bun;
+        var run_in_bun = ctx.debug.run_in_bun or cli.run_in_bun;
```
Comment (telekid) [inline src/cli/bunx_command.zig:156]: Is this logic correct?

File: src/cli/bunx_command.zig [hunk 10]
```diff
@@ -148,43 +149,87 @@ pub const BunxCommand = struct {
...
         var passthrough_list = try std.ArrayList(string).initCapacity(ctx.allocator, argv.len);
+
+        var bin_name = [1]string{""};
+
         var package_name_for_update_request = [1]string{""};
-        {
-            var found_subcommand_name = false;
+
+        if (cli.package.len == 0) {
+            {
```
Comment (telekid) [inline src/cli/bunx_command.zig:165]: self: new fn "parsePackageSpecifiedArgs"

File: src/cli/bunx_command.zig [hunk 11]
```diff
@@ -148,43 +149,87 @@ pub const BunxCommand = struct {
...
+                        package_name_for_update_request[0] = positional;
+                        bin_name[0] = positional;
+                        continue;
+                    }
+
+                    try passthrough_list.append(positional);
+                }
+            }
+        } else {
+            package_name_for_update_request[0] = cli.package;
```
Comment (telekid) [inline src/cli/bunx_command.zig:191]: self: new fn "parsePackageUnspecifiedArgs"

File: src/cli/bunx_command.zig [hunk 12]
```diff
@@ -220,15 +265,6 @@ pub const BunxCommand = struct {
...
 
-        const initial_bin_name = if (strings.eqlComptime(update_request.name, "typescript"))
-            "tsc"
-        else if (update_request.version.tag == .github)
-            update_request.version.value.github.repo.slice(update_request.version_buf)
-        else if (strings.lastIndexOfChar(update_request.name, '/')) |index|
-            update_request.name[index + 1 ..]
-        else
-            update_request.name;
-
```
Comment (telekid) [inline src/cli/bunx_command.zig:231]: Reviewer: I understand the `tsc` special case and believe I have accounted for that, but I don't understand the `(update_request.version.tag == .github)` and `(strings.lastIndexOfChar(update_request.name, '/'))`. 

Specifically, it isn't clear to me why the `bin` name (not that `package` name) would ever be other than the value contained with `argv`.

File: src/cli/bunx_command.zig [hunk 13]
```diff
@@ -148,43 +149,87 @@ pub const BunxCommand = struct {
...
 
         var passthrough_list = try std.ArrayList(string).initCapacity(ctx.allocator, argv.len);
+
+        var bin_name = [1]string{""};
+
         var package_name_for_update_request = [1]string{""};
-        {
-            var found_subcommand_name = false;
+
+        if (cli.package.len == 0) {
```
Comment (telekid) [inline src/cli/bunx_command.zig:164]: The logic in this conditional is confusing. It should really be broken up into something closer to this:

```
if explict_package_mode:
  package_name, bin_name, passthrough_args = parse_argv_explicitly(...)
else
  package_name, bin_name, passthrough_args = parse_argv_implicitly(...)
```

File: src/cli/bunx_command.zig [hunk 14]
```diff
@@ -148,43 +149,87 @@ pub const BunxCommand = struct {
...
                 }
 
-                try passthrough_list.append(positional);
+                if (captured_bin_name and parsing_passthroughs) {
+                    try passthrough_list.append(positional);
+                }
             }
         }
 
+        if (cli.package.len != 0 and bin_name[0].len == 0) {
```
Comment (telekid) [inline src/cli/bunx_command.zig:221]: TODO: Add a test for this exit

File: src/cli/bunx_command.zig [hunk 15]
```diff
@@ -148,43 +149,87 @@ pub const BunxCommand = struct {
...
+            const text =
+                \\ <r><red>error<r><d>:<r> --package specified without indicating command<r>
+                \\ Perhaps you meant the following?:
+                \\ bunx --package {s} <b>-- {s}<r>
+                \\
+            ;
+
+            Output.pretty(text, .{ package_name_for_update_request[0], package_name_for_update_request[0] });
+            Global.exit(1);
```
Comment (telekid) [inline src/cli/bunx_command.zig:230]: Is exit(1) appropriate here?

File: src/cli/bunx_command.zig [hunk 16]
```diff
@@ -148,43 +149,87 @@ pub const BunxCommand = struct {
...
+
+        const cli = try CommandLineArguments.parse(ctx.allocator);
+
         var requests_buf = bun.PackageManager.UpdateRequest.Array.init(0) catch unreachable;
-        var run_in_bun = ctx.debug.run_in_bun;
+        var run_in_bun = ctx.debug.run_in_bun or cli.run_in_bun;
 
         var passthrough_list = try std.ArrayList(string).initCapacity(ctx.allocator, argv.len);
+
+        var bin_name = [1]string{""};
```
Comment (telekid) [inline src/cli/bunx_command.zig:160]: Is this okay? Seems weird but the strategy was already in use for `package_name_for_update_request`.
